LCLSGeneral/LCLSGeneral/POUs/Diagnostics/FB_EtherCATDiag.TcPOU

```diff
@@ -0,0 +1,575 @@
+END_IF
+
+(* acyclic diag *)
+CASE iState OF
+0: (* IDLE *)
+	IF bDiagReq THEN				// diag requested
+		bDiagReq := FALSE;
+
+		IF sMasterNetID <> '' AND sMasterNetID <> '0.0.0.0.0.0' THEN
+			iState := iState + 1;	// execute diag
+			bBusy := TRUE;
+		ELSE
+			bError := TRUE;
+			iErrorID := 7;
+		END_IF
+
+		bEtherCATOK := FALSE;
+	ELSE
+		// check for changes in idle
+		IF (bSlaveCountError OR bMasterDevStateError OR NOT bAllFrameWcStatesOK) AND NOT (arrSlaveInfo[iDiagIndex[0]].bDiagData) THEN
+			bEtherCATOK := FALSE;
+			bDiagReq := TRUE;//new error --> diag requested    
+		ELSIF (bSlaveCountError AND NOT bSlaveCountErrorPrev) OR (NOT bSlaveCountError AND bSlaveCountErrorPrev) THEN
+			bSlaveCountErrorPrev := bSlaveCountError;
+			bEtherCATOK := FALSE;
+			bDiagReq := TRUE;// slave count error change detected --> diag required
+		ELSIF (nMasterDevState <> nMasterDevStatePrev) THEN
+			bEtherCATOK := FALSE;
+			bDiagReq := TRUE;// devstate change detected --> diag required
+		ELSIF (bAllFrameWcStatesOK AND NOT bAllFrameWcStatesOKPrev) OR (NOT bAllFrameWcStatesOK AND bAllFrameWcStatesOKPrev) THEN
+			bAllFrameWcStatesOKPrev := bAllFrameWcStatesOK;
+			bEtherCATOK := FALSE;
+			bDiagReq := TRUE;// frame error change detected --> diag required
+		ELSIF (bSlaveCountError OR bMasterDevStateError OR NOT bAllFrameWcStatesOK OR arrSlaveInfo[iDiagIndex[0]].bDiagData) THEN
+			bEtherCATOK := FALSE;
+		ELSE
+			bEtherCATOK := TRUE;
+		END_IF
+	END_IF
+
+1: (* get adresses *)
+	M_GetSlaveAdresses();
+
+2: (* get states *)
+	M_GetSlaveStates();
+
+3:	(* get topology data length *)
+	M_GetTopoDataLen();
+
+4:	(* get topology data *)
+	M_GetTopoData();
+
+5:	(* scan slaves *)
+	M_ScanSlaves();
+
+6:	(* get identity *)
+	M_GetSlaveIdentity();
+
+7:	(* get name *)
+	M_GetSlaveName();
+
+8:	(* get scanned name *)
+	M_GetScannedSlaveName();
+
+10: (* Log diagnostics *)
+    (* I can't get the fbJsonDataType to actually convert the slave info 
+    structures. I just get nulls. Either I am doing this wrong, or when 
+    the symbol parser encounters datatypes it can't deal with it nulls
+    the whole thing. I'll keep this code commented out until someone figures
+    out how to deal with parsing the slave structs into the json payload *)
+    // DiagnosticJson();
+    // fbLogger(sMsg:='EtherCAT Diagnostic complete', eSevr:=TcEventSeverity.Info);
+    iState := iState + 1;    
+    
+11:	(* DONE *)
+	bBusy := FALSE;
+	iState := 0;
+
+END_CASE
+
+// Log messages
+tEtherCATOK(CLK:=bEtherCATOK);
+IF tEtherCATOK.Q THEN
+    fbLogger(sMsg:='EtherCAT failure, starting diagnostic', eSevr:=TcEventSeverity.Critical, sJson:='');
+END_IF
+
+tFrameWcStateError(CLK:=bFrameWcStateError);
+IF tFrameWcStateError.Q THEN
+    fbLogger(sMsg:='Working Counter Frame Error: error in at least one frame', eSevr:=TcEventSeverity.Error, sJson:='');
+END_IF
+
+tMasterError(CLK:=bMasterDevStateError);
+IF tMasterError.Q THEN
+    fbJson.StartObject();
```

Comment:
It's not as nice as using the structure directly, but I think this isn't so bad either.